pkg/scalers/gitlab_runner_scaler.go

```diff
@@ -0,0 +1,222 @@
+package scalers
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/go-logr/logr"
+	v2 "k8s.io/api/autoscaling/v2"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+
+	"github.com/kedacore/keda/v2/pkg/scalers/scalersconfig"
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+)
+
+const (
+	// externalMetricType is the type of the external metric.
+	defaultTargetPipelineQueueLength = 1
+	// defaultGitlabAPIURL is the default GitLab API base URL.
+	defaultGitlabAPIURL = "https://gitlab.com"
+
+	// pipelineWaitingForResourceStatus is the status of the pipelines that are waiting for resources.
+	pipelineWaitingForResourceStatus = "waiting_for_resource"
+
+	// maxGitlabAPIPageCount is the maximum number of pages to query for pipelines.
+	maxGitlabAPIPageCount = 50
+	// gitlabAPIPerPage is the number of pipelines to query per page.
+	gitlabAPIPerPage = "200"
+)
+
+type gitlabRunnerScaler struct {
+	metricType v2.MetricTargetType
+	metadata   *gitlabRunnerMetadata
+	httpClient *http.Client
+	logger     logr.Logger
+}
+
+type gitlabRunnerMetadata struct {
+	gitlabAPIURL        *url.URL
+	personalAccessToken string
+	projectID           string
+
+	targetPipelineQueueLength int64
+	triggerIndex              int
+}
+
+// NewGitLabRunnerScaler creates a new GitLab Runner Scaler
+func NewGitLabRunnerScaler(config *scalersconfig.ScalerConfig) (Scaler, error) {
+	httpClient := kedautil.CreateHTTPClient(config.GlobalHTTPTimeout, false)
+
+	metricType, err := GetMetricTargetType(config)
+	if err != nil {
+		return nil, fmt.Errorf("error getting scaler metric type: %w", err)
+	}
+
+	meta, err := parseGitLabRunnerMetadata(config)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing GitLab Runner metadata: %w", err)
+	}
+
+	return &gitlabRunnerScaler{
+		metricType: metricType,
+		metadata:   meta,
+		httpClient: httpClient,
+		logger:     InitializeLogger(config, "gitlab_runner_scaler"),
+	}, nil
+}
+
+func parseGitLabRunnerMetadata(config *scalersconfig.ScalerConfig) (*gitlabRunnerMetadata, error) {
```

Comment:
Could you please use the new declarative approach for the metadata parsing?

https://github.com/kedacore/keda/issues/5797

Reply:
this declarative approach is so much more convenient! thanks for pointing it out, I'll fix it.